lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -1,80 +1,196 @@
- * TokenEncoder can then be used to encode and decode tokens.
- */
-class TokenEncoder {
-
-    private final Map<ImmutableByteArray, Integer> decodedToEncoded = new HashMap<>();
-
-    /**
-     * Creates a new TokenEncoder with the given input map. The keys of the map are
-     * the decoded tokens and the values are the encoded tokens.
-     *
-     * @param input the input map
-     */
-    public TokenEncoder(Map<ImmutableByteArray, Integer> input) {
-        this(input, Function.identity());
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class TokenEncoder {
+    public static final int DUMMY_RANK = Integer.MAX_VALUE;
+    public static final int MAX_RANK = Integer.MAX_VALUE - 1;
+    private final Map<ImmutableByteArray, Integer>[] encoders;
+    private int length = 0;
+
+    public TokenEncoder(Map<byte[], Integer> encoder) {
+        if (!encoder.isEmpty()) {
+            TreeMap<Integer, Map<ImmutableByteArray, Integer>> tempEncoders = new TreeMap<>();
+            encoder.forEach((k, v) -> {
+                length++;
+                ImmutableByteArray key = ImmutableByteArray.from(k);
+                tempEncoders.computeIfAbsent(k.length, integer -> new ConcurrentHashMap<>()).put(key, v);
+            });
+            //noinspection unchecked
+            encoders = new ConcurrentHashMap[tempEncoders.lastKey() + 1];
+            tempEncoders.forEach((k, v) -> encoders[k] = v);
+        } else {
+            encoders = new HashMap[0]; // for testing
+        }
+    }
+
+    public static int getMinRankIndex(List<Integer> ranks) {
+        int minRankIndex = -1;
+        int minRank = MAX_RANK;
+
+        int i = 0;
+        int length = ranks.size() - 3;
+        for (; i < length - 2; i += 4) { // Unrolled loop
+            {
+                int r = ranks.get(i);
+                if (r < minRank) {
+                    minRankIndex = i;
+                    minRank = r;
+                }
+            }
+            {
+                int r = ranks.get(i + 1);
+                if (r < minRank) {
+                    minRankIndex = i + 1;
+                    minRank = r;
+                }
+            }
+            {
+                int r = ranks.get(i + 2);
+                if (r < minRank) {
+                    minRankIndex = i + 2;
+                    minRank = r;
+                }
+            }
+            {
+                int r = ranks.get(i + 3);
+                if (r < minRank) {
+                    minRankIndex = i + 3;
+                    minRank = r;
+                }
+            }
+        }
+
+        for (; i <= length; i++) {
+            int r = ranks.get(i);
+            if (r < minRank) {
+                minRankIndex = i;
+                minRank = r;
+            }
+        }
+
+        return minRankIndex;
+    }
+
+    public static int getNextIndex(List<Integer> ranks, int nextIndex) {
```

Comment:
Since we've replaced minimums with sentinels, previous and next indexes are replaced by iteration